### PR TITLE
polygon function to draw visible area

### PR DESCRIPTION
**Pull request by Felecarpp**
A visibility function to get the polygon of visible area from a list of segments (walls). 
There are three new functions in vector.lua used by visibility function: `intersection`, `alignment` and `polar_lt`
I put a demonstration script, unit test script and performance timer script.

### CODE REVIEW

File: vector.lua [hunk 1]
```diff
@@ -119,11 +148,11 @@ function vector:toPolar()
 end
 
 function vector:len2()
```
Comment (HDictus) [inline vector.lua:150]: best to make a separate pull request for these tweaks to keep changes atomic

**Review by HDictus — COMMENTED** [review]
I very much appreciate the additon of unit tests. However, what isn't clear to me from this is what visibles.polygon  is actually for. The demo does not make it much clearer I'm afraid. Cna you explain it further?

**Comment by HDictus** [comment]
> I very much appreciate the additon of unit tests. However, what isn't clear to me from this is what visibles.polygon is actually for. The demo does not make it much clearer I'm afraid. Cna you explain it further?

Ah I feel silly, it is for calculating line of sight. This seems like something handy, but I'm not sure it belongs in hump rather than in a standalone library. 
Either way I'll give it another review so you can use the feedback whether we merge or you make your own library.

File: tests/visible.lua [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  lu.assertEquals(
+    vis.polygon({
+      0, 1, 3, 1,
+      1, 0, 1, 3,
+      3, 2, 2, 3,
+      3, 3, -1, 3,
+      3, 3, 3, -1,
+      -1, -1, -1, 3,
+      -1, -1, 3, -1
+    }),
```
Comment (HDictus) [inline tests/visible.lua:198]: In tests it is often a good idea to be explicit: so pass vector:zero here and elsewhere you rely on it as default.

File: visible.lua [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+local function comp(a, b, center)
+  if a.start == b.start then
+    if a.stop == nil then return false end
+    if b.stop == nil then return true end
+    return vector.alignment(a.start, a.stop, b.stop) > 0
+  end
+  return vector.polar_lt(a.start, b.start, center)
+end
+
+function visible.polygon(segments, center)
```
Comment (HDictus) [inline visible.lua]: The tricky thing with asking for a Vector object here is that it then is compatible only with Vector and not vector-light. People have their reasons for preferring vector-light. You could make it receive two arguments for center: center_or_x and y. Then you can interperet center_or_x as a vector if y is nil, as the x component of a vector from vector-light if y is not nil.

File: tests/perf_visible.lua [hunk 4]
```diff
@@ -0,0 +1,32 @@
+local visible = require("visible")
```
Comment (HDictus) [inline tests/perf_visible.lua:1]: benchmark_visible.lua would be a more conventional filename for this

File: vector.lua [hunk 5]
```diff
@@ -51,6 +51,35 @@ local function randomDirection(len_min, len_max)
 	                 math.random() * (len_max-len_min) + len_min)
 end
 
+-- < 0 -> counterclockwise
+-- = 0 -> colinear
+-- > 0 -> clockwise
+local function alignment(a, b, c)
```
Comment (HDictus) [inline vector.lua:57]: I'll need a bit more thorough a description of what this funtion represents, for example in the docs. From the name alone its purpose is not obvious.

File: tests/visible.lua [hunk 6]
```diff
@@ -0,0 +1,165 @@
+local lu = require("luaunit")
+local vec = require("vector")
+local vis = require("visible")
+
+
+function testOrientation()
```
Comment (HDictus) [inline tests/visible.lua:7]: Thanks for adding tests. The tests for vector functions should have their own file, tests/vector.lua

File: visible.lua [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+  -- generic visibility function
+  -- return concave (frequently) polygon
+  if center == nil then center = vector.zero end
+  local endpoints = {}
+  do
+    local next_endpoints = {}
+    local missing_stops = {}
+    for i = 1, #segments, 4 do
```
Comment (HDictus) [inline visible.lua]: In each loop here you clearly do a distinct operation. Name them and create local functions for them. it makes the code more modular, easier to build on and easier to read.

File: visible.lua [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  end
+  return vector.polar_lt(a.start, b.start, center)
+end
+
+function visible.polygon(segments, center)
+  -- generic visibility function
+  -- return concave (frequently) polygon
+  if center == nil then center = vector.zero end
+  local endpoints = {}
+  do
```
Comment (HDictus) [inline visible.lua]: This do seems to just form a block for no reason. I didn't actually know you could put a do statement alone without causing a syntaxerror.

File: visible.lua (hunk 7 above)
Comment (HDictus) [reply visible.lua]: For example, you might name this one collect_endpoints

File: visible.lua [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+      end
+      -- segments cut endpoints
+      for j, epj in ipairs(next_endpoints) do
+        if j == i then break end
+        if vector.alignment(epi.start, epi.stop, epj.start)
+            * vector.alignment(epi.start, epi.stop, epj.stop) <= 0 and
+            vector.alignment(epj.start, epj.stop, epi.start)
+            * vector.alignment(epj.start, epj.stop, epi.stop) <= 0 then
+          local intersec =
+          vector.intersection(epi.start, epi.stop, epj.start, epj.stop)
```
Comment (HDictus) [inline visible.lua]: formatting is a bit odd here

File: tests/visible.lua [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+  lu.assertTrue(vec.alignment(vec(0, 0), vec(1, 0), vec(0, -1)) > 0)
+  lu.assertTrue(vec.alignment(vec(0, 0), vec(1, 0), vec(0, 1)) < 0)
+  lu.assertTrue(vec.alignment(vec(0, 0), vec(1, 0), vec(2, 0)) == 0)
+  lu.assertTrue(vec.alignment(vec(0, 0), vec(0, -1), vec(0, 1)) == 0)
+  lu.assertTrue(vec.alignment(vec(0, 1), vec(1, 1), vec(2, 1)) == 0)
+end
+
+function testIntersection()
```
Comment (HDictus) [inline tests/visible.lua:15]: This seems like the kind of function where you will want to test edge cases. does it work when they exactly intersect at a point? What does it return when they do not intersect at all?

**Comment by Felecarpp** [comment]
Thanks for these reviews. I will see to them but not for several months.
